core/src/avm2/value.rs

```diff
@@ -183,60 +204,481 @@ pub fn abc_default_value<'gc>(
+            Value::Number(n) => *n,
+            Value::Unsigned(u) => *u as f64,
+            Value::Integer(i) => *i as f64,
+            Value::String(s) => {
+                let strim = s.trim();
+                if strim.is_empty() {
+                    0.0
+                } else if strim.starts_with("0x") || strim.starts_with("0X") {
+                    let mut n: f64 = 0.0;
+                    for c in strim[2..].chars() {
+                        n *= 16.0;
+                        n += match c {
+                            '0' => 0.0,
+                            '1' => 1.0,
+                            '2' => 2.0,
+                            '3' => 3.0,
+                            '4' => 4.0,
+                            '5' => 5.0,
+                            '6' => 6.0,
+                            '7' => 7.0,
+                            '8' => 8.0,
+                            '9' => 9.0,
+                            'a' | 'A' => 10.0,
+                            'b' | 'B' => 11.0,
+                            'c' | 'C' => 12.0,
+                            'd' | 'D' => 13.0,
+                            'e' | 'E' => 14.0,
+                            'f' | 'F' => 15.0,
+                            _ => return Ok(f64::NAN),
+                        };
+                    }
+
+                    n
+                } else {
+                    let (sign, digits) = if strim.starts_with('+') {
+                        (1.0, &strim[1..])
+                    } else if strim.starts_with('-') {
+                        (-1.0, &strim[1..])
+                    } else {
+                        (1.0, strim)
+                    };
+
+                    if digits == "Infinity" {
+                        return Ok(sign * f64::INFINITY);
+                    }
+
+                    //TODO: This is slightly more permissive than ES3 spec, as
+                    //Rust documentation claims it will accept "inf" as f64
+                    //infinity.
+                    sign * digits.parse().unwrap_or(f64::NAN)
+                }
+            }
+            Value::Object(_) => self
+                .coerce_to_primitive(Some(Hint::Number), activation)?
+                .coerce_to_number(activation)?,
+        })
+    }
+
+    /// Coerce the value to a 32-bit unsigned integer.
+    ///
+    /// This function returns the resulting u32 directly; or a TypeError if the
+    /// value is an `Object` that cannot be converted to a primitive value.
+    ///
+    /// Numerical conversions occur according to ECMA-262 3rd Edition's
+    /// ToUint32 algorithm which appears to match AVM2.
+    pub fn coerce_to_u32(&self, activation: &mut Activation<'_, 'gc, '_>) -> Result<u32, Error> {
+        let number = self.coerce_to_number(activation)?;
+
+        Ok(
+            if number == f64::INFINITY
+                || number == f64::NEG_INFINITY
+                || number.is_nan()
+                || number.abs() == 0.0
```

Comment:
I believe `!number.is_finite()` handles all the weirdo cases, and 0 can use the normal case below.
We also already have `avm1::value::f64_to_wrapping_u32` -- could pull this out into a common util module and use it?